### PR TITLE
logging: ot: Fix extern logsystem itf

### PR DESCRIPTION
**Pull request by tkons**
The pull request consist of two commits: 
- first, adding support for full zephyr's log functionality in the interface function of external subsystem
- second, which uses new methods to enhance OpenThread logs.

Note, that zephyr's utils file could not be used in openthread-core-zephyr-config.h.

Tested using logs with all possible parameters combination.

### CODE REVIEW

File: include/logging/log_core.h [hunk 1]
```diff
@@ -577,8 +585,25 @@ void log_hexdump_sync(struct log_msg_ids src_level, const char *metadata,
...
  *
  * @note This function is intended to be used when porting other log systems.
+ *
+ * @param src_level      Log identification.
+ * @param fmt            String to format.
+ * @param ap             Poiner to arguments list.
+ * @param strdup_action  Manages strdup activity.
+ */
+void log_generic(struct log_msg_ids src_level, const char *fmt, va_list ap,
+					enum log_strdup_action strdup_action);
```
Comment (rlubos) [inline include/logging/log_core.h]: Alignment seems improper here.

File: subsys/net/lib/openthread/platform/logging.c [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
```
Comment (rlubos) [inline subsys/net/lib/openthread/platform/logging.c]: A convention for OT related commits is to prefix the tile with `net: openthread: ...`, please update it while rebasing.

File: subsys/net/lib/openthread/platform/logging.c [hunk 3]
```diff
@@ -1,5 +1,5 @@
...
- * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
Comment (rlubos) [inline subsys/net/lib/openthread/platform/logging.c]: 2018 - 2020

**Review by rlubos — APPROVED** [review]
OpenThread changes look good, but you need a rebase to resolve conflicts.

File: include/logging/log_core.h (hunk 1 above)
Comment (tkons) [reply include/logging/log_core.h]: alligned

File: subsys/net/lib/openthread/platform/logging.c (hunk 2 above)
Comment (tkons) [reply subsys/net/lib/openthread/platform/logging.c:1]: changed commit message

File: subsys/net/lib/openthread/platform/logging.c (hunk 3 above)
Comment (tkons) [reply subsys/net/lib/openthread/platform/logging.c]: updated

**Comment by carlescufi** [comment]
@nordic-krch could you please review this one?

File: include/logging/log_core.h [hunk 4]
```diff
@@ -668,6 +693,118 @@ __syscall void z_log_hexdump_from_user(u32_t src_level_val,
...
+ * @param param   Value of actually processed argument
+ * @param action  Action for strdup operation
+ *
+ * @return	  Duplicated string or not changed param
+ */
+static inline log_arg_t z_log_do_strdup(u32_t msk, u32_t idx, log_arg_t param,
+					enum log_strdup_action action)
+{
+#ifndef CONFIG_LOG_MINIMAL
+	char *log_strdup(const char *str);
```
Comment (nordic-krch) [inline include/logging/log_core.h:750]: i know it is necessary to be like this but it's quite messy. I will try to clean it in the future.

File: include/logging/log_core.h [hunk 5]
```diff
@@ -668,6 +693,118 @@ __syscall void z_log_hexdump_from_user(u32_t src_level_val,
...
+
+	if (msk & (1 << idx)) {
+		const char *str = (const char *)param;
+		/* is_rodata(str) is not checked,
+		 * because log_strdup does it.
+		 * Hence, we will do only optional check
+		 * if already not duplicated.
+		 */
+		if (action == LOG_STRDUP_EXEC
+			|| !log_is_strdup(str)) {
```
Comment (nordic-krch) [inline include/logging/log_core.h:760]: wouldn't it fit in single line?

File: include/logging/log_core.h [hunk 6]
```diff
@@ -668,6 +693,118 @@ __syscall void z_log_hexdump_from_user(u32_t src_level_val,
...
+
+#define __LOG_INTERNAL_VA(is_user_context, _src_level, _str, _valist,	       \
+						_argnum, _strdup_action)       \
+do {									       \
+	if (is_user_context) {						       \
+		log_generic_from_user(_src_level, _str, _valist);	       \
+	} else if (IS_ENABLED(CONFIG_LOG_IMMEDIATE)) {			       \
+		log_generic(_src_level, _str, _valist, _strdup_action);	       \
+	} else {							       \
+		if (_argnum == 0)					       \
```
Comment (nordic-krch) [inline include/logging/log_core.h:776]: use `{` for each if else

File: include/logging/log_core.h [hunk 7]
```diff
@@ -407,6 +407,14 @@ static inline char z_log_minimal_level_to_char(int level)
 extern struct log_source_const_data __log_const_start[0];
 extern struct log_source_const_data __log_const_end[0];
 
+/**@brief Enum with possible actions for strdup operation.
```
Comment (nordic-krch) [inline include/logging/log_core.h:410]: add space before brief. comment can be closed in the same line.

**Review by nordic-krch — COMMENTED** [review]
sorry for the delay. Few mainly code style comments.

File: include/logging/log_core.h [hunk 8]
```diff
@@ -668,6 +693,118 @@ __syscall void z_log_hexdump_from_user(u32_t src_level_val,
...
+		log_generic(_src_level, _str, _valist, _strdup_action);	       \
+	} else {							       \
+		if (_argnum == 0)					       \
+			_LOG_INTERNAL_0(_src_level, _str);		       \
+		else {							       \
+			u32_t mask = (_strdup_action != LOG_STRDUP_SKIP) ?     \
+				z_log_get_s_mask(_str, _argnum)		       \
+				: 0;					       \
+									       \
+			if (_argnum == 1)				       \
```
Comment (jukkar) [inline include/logging/log_core.h:783]: This `if` block does not have {} in it.

File: subsys/logging/log_core.c [hunk 9]
```diff
@@ -395,13 +397,33 @@ void log_generic(struct log_msg_ids src_level, const char *fmt, va_list ap)
...
+				const char *str = (const char *)args[idx];
+
+				/* is_rodata(str) is not checked,
+				 * because log_strdup does it.
+				 * Hence, we will do only optional check
+				 * if already not duplicated.
+				 */
+				if (strdup_action == LOG_STRDUP_EXEC
+				   || !log_is_strdup(str)
+				) {
```
Comment (jukkar) [inline subsys/logging/log_core.c:421]: The ending } should be in previous line

File: include/logging/log_core.h (hunk 4 above)
Comment (tkons) [reply include/logging/log_core.h:750]: The solution would be to create log_core_va.h header, included after log_core.h.

File: include/logging/log_core.h (hunk 5 above)
Comment (tkons) [reply include/logging/log_core.h:760]: ok

**Review by jukkar — APPROVED** [review]

File: include/logging/log_core.h (hunk 7 above)
Comment (nordic-krch) [reply include/logging/log_core.h:419]: still missing space.

Comment (tkons) [reply include/logging/log_core.h]: fixed

**Review by tyrantlizard77 — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]

**Comment by nordic-krch** [comment]
@carlescufi  i think we can merge it.

**Comment by carlescufi** [comment]
@tkons could you please rebase and update this PR?

**Comment by MarekPorwisz** [comment]
@nordic-krch I've rebased and fixed merge conflicts. Please have a look if there is something missing. You are more familiar with those changes than I am.

**Pull request merged**